### PR TITLE
Train Step Compile [2/3] optimizer support

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98461
* #100605

Support compiling a user-provided train_step function, provided
- it includes exactly one call to .backward()
- it includes any relevant optimizer steps, followed by zero_grad() calls

Make Adam optimizer (almost) work

Currently, use faketensor execution for the first optimizer step,
to create correct-shaped fake tensors for optimizer states, and
avoid taking the init path on the second try.

Then we hackily convert the optimizer states to zero tensors.  This is the
worst part of the current impl, as it fails to trace the real optimizer
init and could be incorrect.

Finally, we can wrap the real optimizer state tensors with fake tensors
to perform a real train step trace.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

### CODE REVIEW

File: torch/_dynamo/backends/train_step.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+        """
+        Step 3: Functionalize the resulting flattend graph, producing code with copy_ ops
+                as an epilogue for any inplace/mutating ops such as optimizer update.
+        """
+
+        def retraced_f(*args):
+            return Interpreter(fx_g).run(*args)
+
+        # not really sure why we need inference mode here
```
Comment (mrshenli) [inline torch/_dynamo/backends/train_step.py:93]: I assume this is because the backward ops are already captured in the fx.Graph, so we don't need to build autograd graph for it any more. 

Besides that, opt will update params inplace. So if this is run under grad mode, autograd engine will complain that we are inplace updating a leaf node.

File: torch/_dynamo/backends/train_step.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        def call_without_params(*runtime_args):
+            return functional_fx_g(*params_flat + list(runtime_args))
+
+        return call_without_params
+
+    return _compile_fn
+
+
+train_step_eager = train_step_compiler(lookup_backend("eager"))
+register_backend(name="train_step_eager", compiler_fn=train_step_eager)
```
Comment (mrshenli) [inline torch/_dynamo/backends/train_step.py:111]: Does this mean we can directly register `torch.distributed._spmd.compile` as a backend for Dynamo, as long as the input `GraphModule` can represent the full train step?

File: torch/_dynamo/backends/train_step.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        fake_params_flat = fakeify_inputs(params_flat)
+        full_fake_args = fake_params_flat + fake_inputs
+
+        """
+        Step 2: Create a new graphmodule that accepts parameters and user-inputs
+                as inputs to forward(), instead of accessing parameters as attrs.
+        """
+
+        def functional_call(*args, **kwargs):
+            with stateless._reparametrize_module(
```
Comment (mrshenli) [inline torch/_dynamo/backends/train_step.py:68]: I assume we need to reparameterize optimizer states as well, as soon as we use stateful ones?

Like this: 
https://github.com/pytorch/pytorch/blob/482f87a7bc205d8a51744b382002eb23d4e401f2/torch/distributed/_spmd/api.py#L474

**Review by mrshenli — COMMENTED** [review]
Thanks for adding this to Dynamo!

File: torch/_dynamo/backends/train_step.py (hunk 1 above)
Comment (wconstab) [reply torch/_dynamo/backends/train_step.py:93]: > I assume this is because the backward ops are already captured in the fx.Graph, so we don't need to build autograd graph for it any more.

Yes, this is true for sure.  There isn't a .backwards() call anymore.  (which is why i put in the comment: i thought it probably doesn't matter if we run under grad mode since we aren't actually trying to do anything grad related

> autograd engine will complain that we are inplace updating a leaf node.

This is probably the reason- makes sense now.  cc @bdhirsh

File: torch/_dynamo/backends/train_step.py (hunk 2 above)
Comment (wconstab) [reply torch/_dynamo/backends/train_step.py:111]: well, technically yes we could.  Although we should bikeshed the API a bit when we get past prototype stage..

File: torch/_dynamo/backends/train_step.py (hunk 3 above)
Comment (wconstab) [reply torch/_dynamo/backends/train_step.py:69]: thanks, still fighting with dynamo to get the optimizer into the graph.  But will try to use your hint when i get there.

File: torch/_dynamo/backends/train_step.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+def train_step_compiler(backend_compile_fn):
+    """Note [Train Step Compile]
+
+    Usually, torch.compile() allows graph-breaks and compiles pairs of forward (+backward) by
+    extracting sections of forward from python programs and using AotAutograd to produce corresponding
+    chunks of backwards, tying it back together with an AotFunction.
+
+    Instead, TrainStepCompiler assumes the user compiles a full train_step function complete with calls to
+    .backward(), optimizer step(), and zero_grad().  It additionally requires no graph-breaks.
```
Comment (albanD) [inline torch/_dynamo/backends/train_step.py:53]: Note that we want to check that the user doesn't pass `set_to_none=False` to this function to ensure there is no side effect that we wouldn't replicate in the original function.

File: torch/_dynamo/backends/train_step.py (hunk 1 above)
Comment (bdhirsh) [reply torch/_dynamo/backends/train_step.py:144]: ah yep, that's it :)

File: torch/_dynamo/backends/train_step.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+                assert isinstance(x, torch.Tensor)
+                if x not in already_fake:
+                    # Since we do have duplicate names from dynamo refering to the same tensor,
+                    # ensure that we never make more than one faketensor for a given real tensor!
+                    already_fake[x] = fake_mode.from_tensor(x, static_shapes=False)
+                return already_fake[x]
+
+            return [convert(idx, x) for idx, x in enumerate(flat_args)]
+
+        # problem: if i don't remove duplicates here, stateless.Parametrize gets confused by presence of duplicates
```
Comment (wconstab) [inline torch/_dynamo/backends/train_step.py:92]: @albanD this is something strange i am working around for the moment but wanted some help to do properly.

Basically, dynamo puts 2 names to the same tensor on my graphmodule:
  mod.model.layers[0].weight
  and
  mod.__layers_0_weight (or something like that..)


the `__layers_0_weight` name is the one made by dynamo when tracing a usage of the param, so in my simple case that means all uses of the weight are used via this name, and I am getting away with only fake-ifying this name and leaving `model.*` as real tensors when I do the fx tracing.  It might be the case that dynamo can fully gaurantee this, but i'm not sure if it's globally safe or not.

I wanted to fakify all names instead, but the issue I ran into is that 'reparametrize_module' would never do the equivalent of `remove_duplicate=True` for me and the accessor.swap was returning an 'orig_module_state' that had real tensors for the names like '__layers_0_weight' but had my provided faketensors (provided to reparametrize) for names like `model.layers[0].weight`.  Then after the yield, parametrize would 'restore' these faketensors to replace the actual real tensors on the module.

File: torch/_dynamo/output_graph.py [hunk 6]
```diff
@@ -393,6 +406,44 @@ def update_co_names(self, name):
         if name not in self.code_options["co_names"]:
             self.code_options["co_names"] += (name,)
 
+    @staticmethod
+    def module_has_hooks(mod, only_check_unsupported=False):
```
Comment (wconstab) [inline torch/_dynamo/output_graph.py:410]: oops this got checked in by mistake

File: torch/_dynamo/backends/train_step.py (hunk 5 above)
Comment (albanD) [reply torch/_dynamo/backends/train_step.py:95]: I'm not sure to follow exactly. Duplicated params are properly supported in stateless.functioncal_call(). 
If you're using private functions in there I'm not sure though.

**Comment by SherlockNoMad** [comment]
Disclaimer, I don't fully understand what's going on in this PR, but here's some of my reactions.

1. This implementation is relying on the fact that "existing pytorch's adam optimizer lazily initialize optimizer state in the first .step() call". I wonder if this is a general practice for other optimizer implementations. In theory, optimizer state initialization could also happen in the constructor. For those optimizer, we could directly read real optimizer state tensor from opt.state_dict(). 

2. A lot of out-of-core optimizer is a thin python wrapper over C++ kernel, e.g. apex.FusedAdam, deepspeed.FusedAdam... These optimizer are pretty popular, and I wonder how they will behave with train_step_compile.

3. I think XLA's fullgraph ask is more aligned with export path. If that's the case, I would suggest to position this as an export API, instead of torch.compile like API. 

4. I think people will want some annotations to distinguish optimizer subgraph from the rest, i.e. AWS folks is handling optimizer subgraph separately.  

5. I wonder if people would prefer a functionalized representation for optimizer subgraph. In theory, we can aot_functionalized_export for fw_bw_joint graph, and make_fx (without functionalization) for optimizer. 

6. I wonder if people really need a traced-out optimizer subgraph, or just a few opaque `adam.step` node in the resulting graph. (this is what Sigmoid wants...)

7. If we were to expose the resulting joint graph to export user, we should have a clear annotation on graph inputs/outputs, i.e.  OPT_STATE("moment1", "param1"), OPT_STATE("moment2", "param1")... We should make this compatible with aot_export's annotation scheme.

**Comment by wconstab** [comment]
RE optimizers:
 * (1) @janeyx99 can you comment?
 * (2)  yes, for fused optimizers I have not done anything yet, and I should.  Distcompiler may either want users to not use fused optimizers, or provide some way of expanding them to fused calls that operate on smaller sets of params, so they can still be overlapped.  we should be able to add that handling in.
 
RE fullgraph variants
It seems like there may be different use cases that want different flavors.  I want to get some feedback from end-users which flavors they want.  It would be nice to have one solution that satisfied multiple customers and maybe that's possible with some config flags, or we can convince some users they can use a different flavor than they originally asked.
* (3) as discussed offline, not clear if XLA has 1 use case or multiple.  Some xla users (e.g. those using torch-mlir for export) would likely align best with export.  torch-XLA users might prefer .compile style for training at least.  We should follow up here.
* (4 and 5) I am trying to align with AWS/Google users here.  Will follow up
* (6) Distcompiler (cc @mrshenli) wants expanded optimizer to enable reordering. We can brainstorm an alternative format or possibly deferring the expansion of .step() until inside distcompiler instead of inside train_step compiler.
* (7) noted, not sure if we want to go this route or not but would try to align if so.

**Review by voznesenskym — CHANGES_REQUESTED** [review]
Nice prototype, and I am all for train step compiler.

However, a few higher level larger pieces of feedback:

1) Break up this PR please, it is huge - and this makes it hard to know which parts are undertested. LMK if you want some suggestions on the right breakdown here for easier landing and testing.

2) `_is_train_step_compiler` is a hierarchy violation, and we do not want to fork compiler code on this flag. I wrote somewhere before that "every `if export` is a small death to me", and while dramatic, that sentiment stands for what is a first class citizen of dynamo (export) - this backend is less so, and yet feels even more intrusive. We need to find a way to make this generally work without checking the `_is_train_step_compiler` flag ever, anywhere.

3) You are not properly handling checkpoint/restore invariants in your additions to TC. Furthermore, the concept of supporting only a single backwards is also sus to me, and needs to be visited on its own.

**Comment by janeyx99** [comment]
I also don't know what is going on in this PR, but here are my preliminary thoughts/concerns on the optimizers portion:

1. Regarding state initialization, I'm curious about two things: if we can use fake tensors and propagate those shapes/metadata, what requires the initialization to occur? Orthogonal to that, initializing everything to 0 is certainly sketchy. For example, in SGD, the momentum buffers are initialized to the very first grad, which you won't have access to here, and setting it to 0 would not be the same as in eager. Is the plan to re-initialize them to actual values later? This also has relation to my following point, which is we would not be able to control the init state of any other optimizers.
2. Is the train step compiler only meant to work with a few PyTorch optimizers? I agree with Sherlock's point that people write their own optimizers all the time/used the APEX fused optimizers and we would not have insight into the third_party code with this approach. The fused optimizers are also the most popular in terms of performance so far--I'm not sure about the use cases, but I've seen a lot more fused implementations than not.

**Comment by wconstab** [comment]
@voz, i think i can split this into a chunk that does only forward+backward and the chunk that adds optimizer support.

Do you object to all the places inside dynamo where I give special behavior to the train-step compiler?  I definitionally want to give special behaviors to e.g. backward() and optimizer step, so hopefully you're not objecting to that part even though I think it's arguably not nice to have to maintain multiple behaviors for different backends inside dynamo.

**Comment by wconstab** [comment]
@janeyx99 I plan to let distcompiler and XLA folks try this and see if its useful before investing in full coverage for optimizers.  I think one of two things could happen: trainstep is only used in limited cases and we limit the optimizers it works with.  Or, we need to add support for fused optimizers too- but then we also need to accept that the compiler can't optimize those (make them pass-through).

As for the zero init part, agreed it's really sketchy in this PR.  I didn't worry about it as much here bc I am working on deferred init in a follow up PR that will attempt to correctly trace the optimizer init and then play it back.

**Comment by wconstab** [comment]
@janeyx99 RE 
>Regarding state initialization, I'm curious about two things: if we can use fake tensors and propagate those shapes/metadata, what requires the initialization to occur? Orthogonal to that, initializing everything to 0 is certainly sketchy. For example, in SGD, the momentum buffers are initialized to the very first grad, which you won't have access to here, and setting it to 0 would not be the same as in eager. Is the plan to re-initialize them to actual values later? This also has relation to my following point, which is we would not be able to control the init state of any other optimizers.

The optimizer object will "initialize" itself (lazily, at least in the case of sgd/adam), during faketensor tracing, even if I only cared about getting the output shapes at that time.  It is impossible to trace the logic of .step without first having faketensor objects in place of all the optimizer states which may be referenced during .step.

So one problem is that after fake-tracing, I have an optimizer object that has fake tensors, alongside my nn.module which has real tensors.  I intend to return a callable to the user that works with their module and optimizer objects, but its a bit odd to require them to manually intervene after compilation and real-ify their fake optimizer.

Another problem which i deal with is that if I trace the train_step on the first pass, my traced graph also captures the factory-function calls that the optimizer issues, and I don't want that.  So I first warm it up to ensure it has fake states, then I trace the 'warmed' optimizer, finally make it real.  Maybe i'm wrong here: I could try letting the user own this step.

In the distcompiler use case, the ask is for the model and optimizer to go into the compiler as fake and be initialized after distcompiler does sharding.  It could be that both model/opt get 'realized' inside distcompiler after sharding, or, both get realized by the user after compilation.  cc @mrshenli @fegin for ideas there

**Review by kumpera — COMMENTED** [review]
This looks great!
Do you see this having the same setup for large models as DistCompiler is taking? I.E. models would be initialized on the meta device and have the train step compiler materialize them?

File: test/dynamo/test_train_step.py [hunk 7]
```diff
@@ -53,19 +56,147 @@ def train_step(model, inputs):
...
+
+        correct_params = {
+            name: param.clone().detach() for name, param in model.named_parameters()
+        }
+
+        opt_params = {
+            name: param.clone().detach() for name, param in opt_model.named_parameters()
+        }
+
+        self.assertTrue(same(correct_loss, opt_loss))
```
[the remark this answers is not in the document]
Comment (wconstab) [reply test/dynamo/test_train_step.py:150]: oh, good catch. it was probably bad copypaste

**Comment by wconstab** [comment]
Question for DistCompiler (cc @mrshenli @wanchaol) and XLA (cc @JackCaoG) users about how they want optimizer support to work.
Note: i'm still working separately on how to handle 'deferred initialization' which could totally change how optimizers work.  Let's discuss deferred-init optimizers separately in the next PR.  For this PR, think about the case where the model is real.

The question: what if instead of the hacky approach where I convert the fake optimizer states into zero-initialized tensors after tracing, I just leave them all fake?  This would mean that users of train_step_compiler would have to manually init their optimizer after compiling.  The situation could be improved when we add support for deferred init, but at least for now it would be a more correct but less convenient state for initial users (mainly those doing tire-kicking and giving feedback).

**Comment by wconstab** [comment]
> Today, in the compile function, we look up cache and run that. Is it possible to also init the optimizer there

IIUC you're suggesting that in the dynamo layer where we find a cache-hit for the frame, before we execute the cached thing, we notice there is (are) an optimizer(s) as arguments to the frame, then check if they are initialized, and if not, call an init function?  Also, we provide some way for a user to provide the init function(s) to the _compile_train_step api in the first place, so they can be paired up with the right optimizers and stashed for later?

> if the init API is provided by us, I assume we can do that in the compiled function and make it transparent to users

So, I think the above is possible, but it would just be a lot cleaner if instead the optimizer itself provided the init api.  If the optimizer did not provide .init() for us, then i think the next cleanest thing to do is for the user to stuff .init onto the optimizer before compiling, and then just check for it and call it inside the train-step compiler.  Setting it as an attribute on the optimizer itself removes the need to have other infra inside dynamo to keep track of it.

**Comment by wanchaol** [comment]
Thanks for pushing hard on train step graph capture! My two cents on whether we should expose the optimizer init: I do think we should do this, today there's a limitation that simply disallow us to either do train step capture, or do functorch on the existing pytorch optimizers. I think for most of the OSS world people uses standard pytorch optimizers (if the optimizer/optimization algorithm becomes very popular, we usually consider implement or upstream the optimizer to core), even if we enable pytorch standard optimizer with state init and allow PT2 to capture it, that would be a great win already.

In regards to the fact that there're all kinds of third-party optimizers that @SherlockNoMad and @janeyx99 mentioned, I totally agree on that and this is indeed a issue, the fact is that we can control nothing about what they did. But I think it's very similar to the situation with the general PT2 capture, we can't control what user wrote their custom nn.Module too and it might contain data dependent control flows or thin python wrapper C++ kernel in their forward function, so in reality we should simply graph break for those cases. I think if we introduce the optimizer state init API with a BC compatible way, and for those third party optimizers without it, optimizer capture in train step would graph break, it would resolve the usability concern and allow optimizer authors to gradually move to this `init` API when they want to utilize PT2 for their optimizers.
